### ISSUE TITLE
Private methods of base classes somehow don't wanna be compiled

### ISSUE THREAD

**Issue by k06a** [ethereum/solidity#9610]
## Description

Private methods of the same signature (name + args) gives compilation error when inherited from both. I expect private methods of the contract should not produce any collisions on the inherited contracts.

## Environment

- Compiler version: 0.6.12

## Steps to Reproduce

```solidity
pragma solidity ^0.6.0;

contract A {
    function _f() private pure {
    }
}

contract B {
    function _f() private pure {
    }
}

contract C is A, B {
}
```
-->
```
browser/1_Storage.sol:13:1: TypeError: Derived contract must override function "_f". Two or more base classes define function with same name and parameter types.
contract C is A, B {
^ (Relevant source part starts here and spans across multiple lines).
browser/1_Storage.sol:4:5: Definition in "A":
function _f() private pure {
^ (Relevant source part starts here and spans across multiple lines).
browser/1_Storage.sol:9:5: Definition in "B":
function _f() private pure {
^ (Relevant source part starts here and spans across multiple lines).
```

**Comment by chriseth** [comment]
"private" currently does not mean that such functions are invisible - they are only inaccessible. The reason is that it could easily lead to confusion if someone misses the fact that one function is private.

**Comment by chriseth** [comment]
Is making private functions invisible a feature you strongly rely or or is it just something you find unusual?

**Comment by k06a** [comment]
@chriseth but if I would like to inherit from 2 smart contracts from different developers their private methods could produce a collision and compilation error. It feels like a bug.

**Comment by chriseth** [comment]
We are currently trying to add features to the language that make inheritance much less important - maybe that could be a solution. Can you show a specific example?

**Comment by gitpushOmnik** [comment]
I think this might count as an example related to k06a's problem.

File1 : contract1.sol

    pragma solidity >=0.5.0 <0.7.0;
    contract Base1 {
        uint a=4;
        uint b=6;
        function pr() private view returns(uint){
            return a+b;
        }
    }


File2: contract2.sol

    pragma solidity >=0.5.0 <0.7.0;
    contract Base2  {
        uint c=7;
        uint d=8;
        function pr() private view returns(uint){
           return c+d;
        }
    }

Main.sol

    pragma solidity >=0.5.0 <0.7.0;
    import "contract1.sol";
    import "contract2.sol";
    contract derived is Base1,Base2 {
        function ca() public view returns(uint){return Base1.pr();}
    }


After compiling Main.sol, we get the following error : 

    Derived contract must override function "pr". Two or more base classes define function with same name and parameter 
    types. contract derived is Base1,Base2 { ^ (Relevant source part starts here and spans across multiple lines). 
    contract1.sol:5:9: Definition in "Base1": function pr() public view returns(uint){ ^ (Relevant source part starts here and 
    spans across multiple lines). contract2.sol:5:9: Definition in "Base2": function pr() private view returns(uint){ ^ (Relevant 
    source part starts here and spans across multiple lines).

I think the problem here is that Solidity doesn't accept multiple inheritance with private base class functions having the same name.It throws a TypeError. But if you look through a broader perspective, then private functions of Base classes are not much use directly to derived contracts. 

Nevertheless, it is not the case for private functions only. Even public functions cause the same ambiguity.And derived functions use public base classes functions. 

File1 : contract1.sol

    pragma solidity >=0.5.0 <0.7.0;
    contract Base1 {
        uint a=4;
        uint b=6;
        function pu() public view returns(uint){
            return a+b;
        }
    }


File2: contract2.sol

    pragma solidity >=0.5.0 <0.7.0;
    contract Base2  {
        uint c=7;
        uint d=8;
        function pu() public view returns(uint){
           return c+d;
        }
    }

Main.sol

    pragma solidity >=0.5.0 <0.7.0;
    import "contract1.sol";
    import "contract2.sol";
    contract derived is Base1,Base2 {
        function ca() public view returns(uint){return Base1.pu();}
    }

    browser/main.sol:6:5: TypeError: Derived contract must override function "pu". Two or more base classes define function 
    with same name and parameter types. contract derived is Base1,Base2 { ^ (Relevant source part starts here and spans 
    across multiple lines). contract1.sol:5:9: Definition in "Base1": function pu() public view returns(uint){ ^ (Relevant source 
    part starts here and spans across multiple lines). contract2.sol:5:9: Definition in "Base2": function pu() public view 
    returns(uint){ ^ (Relevant source part starts here and spans across multiple lines).

In a nutshell, Solidity doesn't allow multiple inheritance with base classes functions having same name, regardless of whether they're public or private. 
Should I open up a new issue for this?

**Comment by k06a** [comment]
Any progress on this? I believe private methods should not be exposed in any form, even to produce a compile-time errors because of the "same" naming.

**Comment by ekpyron** [comment]
> Any progress on this? I believe private methods should not be exposed in any form, even to produce a compile-time errors because of the "same" naming.

Unfortunately not - I don't see any really *strong* opposition to this in either of the two issues, but there was never an actual decision for or against doing it and then apparently it got lost amidst other priorities - I can try moving it up on our agenda again. EDIT: it's actually already our first item on our design backlog, so hopefully we can make some progress soon.

**Comment by cameel** [comment]
Yeah, I bumped it in design backlog after seeing the comment in the other issue. We can talk about it at the [design call next Wednesday](https://docs.soliditylang.org/en/latest/contributing.html#team-calls). Also, a reminder that the design calls are open in case anyone would like to join. The decision here looks straightforward to me though (I think we should not report an error here) so I don't really expect much discussion to be necessary.

So basically #11889 is a duplicate of this. I think we should keep only one open. Which one though?

**Comment by k06a** [comment]
@cameel mark this as duplicate, since #11889 got more attention

**Comment by ekpyron** [comment]
Closing in favour of https://github.com/ethereum/solidity/issues/11889
